File: tools/browser/audit.js
import {ESPLoader, Transport} from 'esptool-js';
import {createSHA256, md5, sha256} from 'hash-wasm';
import {HARDWARE_PROFILE, PARTITIONS} from './profile.js';

export const FLASH_BYTES = HARDWARE_PROFILE.resources.flash_bytes;
export const BLOCK = 64 * 1024;
export const PROFILE = HARDWARE_PROFILE.installation_id;
export const STOCK = HARDWARE_PROFILE.layout.partitions.map(p=>[p.name,p.type,p.subtype,p.offset,p.size]);
export const hex = bytes => Array.from(bytes, b => b.toString(16).padStart(2,'0')).join('');
export function ensure(value, message, code) {if (!value) {const error=new Error(message);error.userMessage=message;if(code)error.code=code;throw error;}}
const words = (...values) => {
  const bytes = new Uint8Array(values.length*4), view = new DataView(bytes.buffer);
  values.forEach((value,i) => view.setUint32(i*4,value,true)); return bytes;
};
const aborted = signal => {if(signal?.aborted) throw new DOMException('Cancelled','AbortError');};

export async function parseTable(bytes, address=0x8000, capacity=FLASH_BYTES) {
  const entries=[];
  for(let cursor=0; cursor<0xc00; cursor+=32) {
    const entry=bytes.slice(cursor,cursor+32);
    ensure(entry.length===32,'Truncated partition table.');
    const view=new DataView(entry.buffer,entry.byteOffset,entry.byteLength);
    if(view.getUint16(0,true)===0xebeb) {
      ensure(entries.length && entry.slice(2,16).every(b=>b===255) &&
        await md5(bytes.slice(0,cursor))===hex(entry.slice(16)), 'Partition MD5 mismatch.');
      return entries;
    }
    ensure(view.getUint16(0,true)===0x50aa,'Missing partition MD5.');
    const raw=entry.slice(12,28), end=raw.indexOf(0);
    const name=String.fromCharCode(...raw.slice(0,end<0?16:end));
    const p={name,type:entry[2],subtype:entry[3],offset:view.getUint32(4,true),size:view.getUint32(8,true),flags:view.getUint32(28,true)};
    ensure(/^[\x20-\x7e]{1,16}$/.test(name),'Invalid partition name.');
    ensure(p.size && !(p.offset%4096) && !(p.size%4096) && p.offset>=address+4096 &&
      p.offset+p.size<=capacity && (p.type!==0 || !(p.offset%65536)), 'Partition exceeds flash or is unaligned.');
    ensure(!entries.some(q=>q.name===name || p.offset<q.offset+q.size && q.offset<p.offset+p.size),'Overlapping or duplicate partition.');
    entries.push(p);
  }
  throw new Error('Missing partition MD5.');
}

export function matchesStock(partitions) {
  return partitions.length===STOCK.length && STOCK.every(([name,type,subtype,offset,size],i)=>{
    const p=partitions[i]; return p.name===name && p.type===type && p.subtype===subtype && p.offset===offset && p.size===size && p.flags===0;
  });
}

export async function inspectImage(bytes) {
  ensure(bytes.length>=24 && bytes[0]===0xe9 && bytes[1]>0 && bytes[1]<=16,'Unknown application/bootloader image.');
  const view=new DataView(bytes.buffer,bytes.byteOffset,bytes.byteLength);
  ensure(view.getUint16(12,true)===18 && bytes[23]===1,'Expected a digested ESP32-P4 image.');
  let cursor=24,checksum=0xef;
  for(let i=0;i<bytes[1];i++) {
    ensure(cursor+8<=bytes.length,'Truncated image segment header.');
    const length=view.getUint32(cursor+4,true);cursor+=8;
    ensure(length<=bytes.length-cursor,'Image segment exceeds partition.');
    for(let j=cursor;j<cursor+length;j++)checksum^=bytes[j];cursor+=length;
  }
  const checksumAt=Math.floor(cursor/16)*16+15,end=checksumAt+1;
  ensure(end+32<=bytes.length && bytes[checksumAt]===checksum,'Image checksum mismatch.');
  ensure(await sha256(bytes.slice(0,end))===hex(bytes.slice(end,end+32)),'Image SHA-256 mismatch.');
  const result={image_bytes:end+32,chip_id:18,internal_checksum_verified:true,appended_sha256_verified:true,
    min_revision:view.getUint16(15,true),max_revision:view.getUint16(17,true)};
  if(bytes.length>=288 && view.getUint32(32,true)===0xabcd5432) {
    const field=(offset,length)=>new TextDecoder().decode(bytes.slice(32+offset,32+offset+length)).split('\0')[0].replace(/[^\x20-\x7e]/g,'?');
    result.application={version:field(16,32),project:field(48,32),idf:field(112,32)};
  }
  return result;
}

export async function analyzeBackup(file, progress=()=>{}, signal) {
  ensure(file.size===FLASH_BYTES,'A complete 32 MiB backup is required.');
  const hash=await createSHA256(); hash.init();
  const tables=[];
  let erased=true;
  for(let offset=0;offset<file.size;offset+=BLOCK) {
    aborted(signal);
    const bytes=new Uint8Array(await file.slice(offset,offset+BLOCK).arrayBuffer());
    hash.update(bytes);
    for(let at=0;at<bytes.length;at+=4096) {
      if(bytes[at]===0xaa && bytes[at+1]===0x50) {
        try {tables.push({offset:offset+at,partitions:await parseTable(bytes.slice(at,at+4096),offset+at)});} catch { /* Not a valid table candidate. */ }
      }
    }
    const slot=PARTITIONS[HARDWARE_PROFILE.layout.initial_app];
    const start=Math.max(0,slot.offset-offset),end=Math.min(bytes.length,slot.offset+slot.size-offset);
    if(end>start && !bytes.slice(start,end).every(b=>b===255)) erased=false;
    progress(offset+bytes.length,file.size);
  }
  ensure(tables.length===1,'Expected one unambiguous MD5-verified partition table.');
  const table=tables[0];
  const images=[];
  // Only the reviewed table location establishes the bounded bootloader region here.
  if(table.offset===0x8000) {
    aborted(signal);
    const boot=new Uint8Array(await file.slice(0x2000,0x8000).arrayBuffer());
    images.push({name:'bootloader',offset:0x2000,region_sha256:await sha256(boot),...await inspectImage(boot)});
  }
  for(const p of table.partitions.filter(p=>p.type===0)) {
    aborted(signal);
    const bytes=new Uint8Array(await file.slice(p.offset,p.offset+p.size).arrayBuffer());
    if(bytes.every(b=>b===255)){images.push({name:p.name,erased:true});continue;}
    images.push({name:p.name,offset:p.offset,...await inspectImage(bytes)});
  }
  const table_sha256=await sha256(new Uint8Array(await file.slice(table.offset,table.offset+4096).arrayBuffer()));
  return {schema:1,sha256:hash.digest('hex'),flash_bytes:file.size,partition_table_offset:table.offset,
    partition_table_md5_verified:true,partitions:table.partitions,images,table_sha256,
    stock_layout_matches:table.offset===0x8000 && matchesStock(table.partitions),
    ota_1_erased:erased,installable:false};
}

export function decodeSecurity(bytes) {
  ensure(bytes instanceof Uint8Array && bytes.length===20,'Security state unavailable.','security_unverified');
  const v=new DataView(bytes.buffer,bytes.byteOffset,bytes.byteLength),flags=v.getUint32(0,true);
  ensure(v.getUint32(12,true)===18 && !(flags&7) && bytes[4]===0,'Unsupported chip or enabled security.','security_unsupported');
  return {flash_crypt_cnt:bytes[4],parsed_flags:{SECURE_BOOT_EN:false,SECURE_BOOT_AGGRESSIVE_REVOKE:false,SECURE_DOWNLOAD_ENABLE:false}};
}

export async function openReader(port, {Loader=ESPLoader,SerialTransport=Transport,baud=460800}={}) {
  const transport=new SerialTransport(port,false);
  const loader=new Loader({transport,baudrate:baud,debugLogging:false,terminal:{clean(){},write(){},writeLine(){}}});
  try {
    await loader.connect('default_reset',3,true);
    const revision=await loader.chip.getChipRevision(loader);
    ensure(loader.chip.CHIP_NAME===HARDWARE_PROFILE.chip.name && revision>=HARDWARE_PROFILE.chip.revision_min &&
      revision<=HARDWARE_PROFILE.chip.revision_max,'Expected ESP32-P4 revision 1.3.','chip_revision_mismatch');
    // GET_SECURITY_INFO is an Espressif ROM command; parse the pinned esptool 5.4.0 format.
    const security=decodeSecurity(await loader.checkCommand('read security',0x14,new Uint8Array(),0,20));
    // ROM SPI_ATTACH includes its reserved legacy word; no flash_begin is used to attach.
    await loader.checkCommand('attach flash',loader.ESP_SPI_ATTACH,words(0,0));
    const flashId=await loader.readFlashId();
    ensure((flashId>>>16 & 255)===25,'Expected a 32 MiB flash chip.','flash_capacity_mismatch');
    const identity=await loader.chip.readMac(loader); // Private RAM/file only, never an ownership proof.
    if(loader.chip.postConnect) await loader.chip.postConnect(loader);
    await loader.runStub();
    await loader.flashSpiAttach(0);
    if(baud!==115200) await loader.changeBaud();
    return {loader,transport,hardware:{chip:'ESP32-P4',revision:103,flash_bytes:FLASH_BYTES,security,identity,reader:'esptool-js-0.6.1',baud}};
  } catch(error) {await transport.disconnect();throw error;}
}

// The pinned esptool-js readFlash omits the terminal MD5 frame and accumulates by copying.
// Use its transport/command implementation with Espressif's documented stub READ_FLASH
// framing, bounded buffers and the digest check used by Python esptool 5.4.0.
export async function readChunk(reader, offset, size, signal, onProgress=()=>{}) {
  aborted(signal);
  ensure(size>0 && size<=BLOCK && !(size%4096),'Invalid bounded read.');
  await reader.loader.checkCommand('read flash',reader.loader.ESP_READ_FLASH,words(offset,size,4096,64));
  const output=new Uint8Array(size); let received=0;
  while(received<size) {
    aborted(signal);
    const packet=await reader.transport.read(5000);
    ensure(packet instanceof Uint8Array && packet.length===Math.min(4096,size-received),'Incomplete flash packet; retry this read.');
    output.set(packet,received); received+=packet.length;
    await reader.transport.write(words(received)); onProgress(received);
  }
  const digest=await reader.transport.read(5000);
  ensure(digest instanceof Uint8Array && digest.length===16 && hex(digest)===await md5(output),'Flash transfer MD5 mismatch.');
  return output;
}

export async function captureRead(reader, handle, signal, progress) {
  const writer=await handle.createWritable(); const hash=await createSHA256();hash.init();
  try {
    for(let offset=0;offset<FLASH_BYTES;offset+=BLOCK) {
      const bytes=await readChunk(reader,offset,BLOCK,signal,n=>progress(offset+n,FLASH_BYTES));
      await writer.write(bytes);hash.update(bytes);
    }
    aborted(signal);await writer.close();return hash.digest('hex');
  } catch(error) {await writer.abort().catch(()=>{});throw error;}
}

export async function compareBackups(first,second,progress=()=>{},signal) {
  const a=await analyzeBackup(first,(n,total)=>progress('Verifying backup A',n,total),signal);
  const b=await analyzeBackup(second,(n,total)=>progress('Verifying backup B',n,total),signal);
  ensure(a.sha256===b.sha256,'Backups differ. Keep both privately and repeat the audit.');
  return {...a,matching_files:true,independent_reads_match:false,
    evidence:'Imported files match; independent physical reads require a capture record or owner evidence.'};
}

export {sha256};

File: firmware/tools/audit.py
"""Read-only ESP32-P4 audit and private backup inspection. No flash-write commands."""
import argparse
import contextlib
import hashlib
import json
import os
import struct
from pathlib import Path

MIB = 1024 * 1024


def digest(data):
    return hashlib.sha256(data).hexdigest()


def image_metadata(data):
    """Bounded offline image verification using the pinned maintained esptool parser."""
    import io
    import esptool
    from esptool.bin_image import ESP32P4FirmwareImage
    if esptool.__version__ != '5.4.0':raise ValueError('Use pinned esptool 5.4.0')
    image=ESP32P4FirmwareImage(io.BytesIO(data))
    if image.chip_id!=18 or not image.append_digest or image.stored_digest!=image.calc_digest or image.checksum!=image.calculate_checksum():
        raise ValueError('Invalid P4 image checksum/digest/header')
    return {'image_bytes':image.data_length+32,'chip_id':image.chip_id,'min_revision':image.min_rev_full,
        'max_revision':image.max_rev_full,'flash_mode':image.flash_mode,'flash_size_frequency':image.flash_size_freq,
        'internal_checksum_verified':True,'appended_sha256_verified':True}


def private_directory(path):
    path = path.resolve()
    if any((parent/'.git').exists() for parent in [path, *path.parents]):
        raise ValueError('Keep private audits and backups outside every Git checkout')
    path.mkdir(mode=0o700, parents=True, exist_ok=False)
    return path


def partition_table(data, address, flash_size):
    sector = data[address:address+4096]
    partitions, entries = [], bytearray()
    for cursor in range(0, min(len(sector), 0xC00), 32):
        entry = sector[cursor:cursor+32]
        if len(entry) != 32:
            raise ValueError('Truncated partition table')
        if entry[:2] == b'\xeb\xeb':
            if entry[2:16] != b'\xff'*14 or hashlib.md5(entries).digest() != entry[16:]:
                raise ValueError('Partition MD5 mismatch')
            if not partitions:
                raise ValueError('Empty partition table')
            return partitions
        magic, kind, subtype, offset, size, label, flags = struct.unpack('<HBBII16sI', entry)
        if magic != 0x50AA:
            raise ValueError('Missing partition MD5 or invalid entry')
        name = label.split(b'\0', 1)[0].decode('ascii')
        if not name or any(ord(c) < 32 or ord(c) > 126 for c in name):
            raise ValueError('Invalid partition name')
        if not size or offset % 4096 or size % 4096 or offset < address+4096 or offset+size > flash_size:
            raise ValueError('Partition outside flash or unaligned')
        if kind == 0 and offset % 65536:
            raise ValueError('App partition is not 64 KiB aligned')
        if any(offset < p['offset']+p['size'] and p['offset'] < offset+size for p in partitions):
            raise ValueError('Overlapping partitions')
        if any(name == p['name'] for p in partitions):
            raise ValueError('Duplicate partition name')
        partitions.append(dict(name=name, type=kind, subtype=subtype, offset=offset, size=size, flags=flags))
        entries.extend(entry)
    raise ValueError('No partition MD5')


def analyze(data):
    size = len(data)
    if size < 2*MIB or size > 64*MIB or size & (size-1):
        raise ValueError('Expected a complete power-of-two flash dump (2–64 MiB)')
    candidates = []
    for address in range(0, size-4095, 4096):
        if data[address:address+2] != b'\xaa\x50':
            continue
        try:
            partitions = partition_table(data, address, size)
        except (ValueError, UnicodeDecodeError):
            continue
        candidates.append((address, partitions))
    if len(candidates) != 1:
        raise ValueError(f'Expected one unambiguous MD5-verified partition table; found {len(candidates)}')
    address, partitions = candidates[0]
    for partition in partitions:
        start, length = partition['offset'], partition['size']
        partition['sha256'] = digest(data[start:start+length])
        if partition['type'] == 0:
            # ESP image header (24), first segment header (8), then app descriptor.
            desc = data[start+32:start+288]
            if data[start] == 0xE9 and desc[:4] == b'\x32\x54\xcd\xab':
                def field(offset, count):
                    raw = desc[offset:offset+count].split(b'\0', 1)[0]
                    return ''.join(chr(c) if 32 <= c <= 126 else '?' for c in raw)
                partition['app_metadata'] = {'version': field(16, 32), 'project': field(48, 32), 'idf': field(112, 32)}
    return {'schema': 1, 'flash_bytes': size, 'sha256': digest(data),
            'partition_table_offset': address, 'partition_table_md5_verified': True,
            'partitions': partitions, 'installable': False,
            'limitations': ['Offline bytes do not prove board identity or security state.',
                           'Partition discovery does not verify the stock bootloader address or restore procedure.']}


def security_is_unprotected(info):
    flags = info.get('parsed_flags', {})
    return (type(info.get('flash_crypt_cnt')) is int and info['flash_crypt_cnt'] == 0
            and all(flags.get(key) is False for key in
                    ['SECURE_BOOT_EN', 'SECURE_DOWNLOAD_ENABLE', 'SECURE_BOOT_AGGRESSIVE_REVOKE']))


def capture(port, output, baud, use_ram_stub):
    import esptool
    from esptool import cmds
    if esptool.__version__ != '5.4.0':
        raise ValueError('Use the pinned esptool 5.4.0 environment')
    output = private_directory(output)
    esp = None
    # esptool may print identifiers; keep all diagnostics in the private directory.
    with (output/'serial-private.log').open('w') as log, contextlib.redirect_stdout(log), contextlib.redirect_stderr(log):
        try:
            esp = cmds.detect_chip(port=port, baud=115200)
            if esp.CHIP_NAME != 'ESP32-P4' or esp.get_chip_revision() != 103:
                raise ValueError('Expected ESP32-P4 revision 1.3; stop and review this unit')
            info = esp.get_security_info()
            if not security_is_unprotected(info):
                raise ValueError('Security state enabled or unknown; stop for a separate recovery review')
            cmds.attach_flash(esp)
            flash_id = esp.flash_id()
            capacity = (flash_id >> 16) & 255
            if capacity != 25:
                raise ValueError('Expected 32 MiB JEDEC capacity; do not guess a flash size')
            metadata = {'schema': 1, 'chip': esp.CHIP_NAME, 'revision': 103,
                        'flash_bytes': 1 << capacity, 'jedec_id': flash_id, 'security': info,
                        'ram_stub': use_ram_stub, 'esptool': esptool.__version__}
            (output/'hardware-private.json').write_text(json.dumps(metadata, indent=2)+'\n')
            if use_ram_stub:
                esp = cmds.run_stub(esp)
                cmds.attach_flash(esp)
            if baud != 115200:
                esp.change_baud(baud)
            hashes = []
            for name in ['backup-a.bin', 'backup-b.bin']:
                cmds.read_flash(esp, 0, 1 << capacity, str(output/name), flash_size='32MB', no_progress=True)
                data = (output/name).read_bytes()
                if len(data) != 1 << capacity:
                    raise ValueError('Incomplete backup; never use it for restoration')
                hashes.append(digest(data))
            if hashes[0] != hashes[1]:
                raise ValueError('Independent flash reads differ; neither is approved for restoration')
            report = analyze(data)
            report['independent_reads_match'] = True
            report['hardware'] = metadata
            (output/'audit-private.json').write_text(json.dumps(report, indent=2)+'\n')
        except Exception:
            import traceback
            traceback.print_exc()
            raise
        finally:
            if esp is not None:
                esp._port.close()
            for path in output.iterdir():
                path.chmod(0o600)
    print('Two matching reads and an offline audit saved privately. Nothing flashed. Reset the board manually.')
    print('Keep another copy on separate storage. This report does not authorize installation.')


def main():
    os.umask(0o077)
    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest='command', required=True)
    commands.add_parser('ports')
    inspect = commands.add_parser('inspect'); inspect.add_argument('backup', type=Path)
    grab = commands.add_parser('capture'); grab.add_argument('--port', required=True)
    grab.add_argument('--output', required=True, type=Path)
    grab.add_argument('--baud', choices=[115200, 460800], type=int, default=115200)
    grab.add_argument('--ram-stub', action='store_true', help='Opt in to a temporary RAM reader; no flash writes')
    args = parser.parse_args()
    if args.command == 'ports':
        from serial.tools.list_ports import comports
        for port in comports():
            print(json.dumps({'port': port.device, 'vid': port.vid, 'pid': port.pid}))
    elif args.command == 'inspect':
        # Contains partition names/app metadata; inspect locally, never upload a flash dump.
        print(json.dumps(analyze(args.backup.read_bytes()), indent=2))
    else:
        capture(args.port, args.output, args.baud, args.ram_stub)


if __name__ == '__main__':
    try:
        main()
    except Exception as error:
        raise SystemExit(f'Audit stopped ({type(error).__name__}). Review the private log locally; no installation approved.')
